### PR TITLE
Updating  to the nodal_concentrated_element

### PR DESCRIPTION
**Pull request by loumalouomega**
Adding checks and corrections

Regarding #1230

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp [hunk 1]
```diff
@@ -495,63 +461,23 @@ int NodalConcentratedElement::Check( const ProcessInfo& rCurrentProcessInfo )
...
-        {
-            KRATOS_ERROR << "Missing variable DISPLACEMENT on node " << this->GetGeometry()[i].Id() << std::endl;
-        }
-
-        if ( this->GetGeometry()[i].HasDofFor( DISPLACEMENT_X ) == false || this->GetGeometry()[i].HasDofFor( DISPLACEMENT_Y ) == false || this->GetGeometry()[i].HasDofFor( DISPLACEMENT_Z ) == false )
-        {
-            KRATOS_ERROR << "Missing one of the dofs for the variable DISPLACEMENT on node " << GetGeometry()[i].Id() << std::endl;
-        }
+    // Verify that the variables are correctly initialized
+    KRATOS_ERROR_IF(VELOCITY.Key() == 0) << "VELOCITY has Key zero! (check if the application is correctly registered" << std::endl;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: Can you use the functions from checks.h?

**Review by philbucher — CHANGES_REQUESTED** [review]

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp [hunk 2]
```diff
@@ -300,36 +279,40 @@ void NodalConcentratedElement::CalculateRightHandSide(VectorType& rRightHandSide
...
-        Volume_Acceleration = GetGeometry()[0].FastGetSolutionStepValue(VOLUME_ACCELERATION);
-    }
+        volume_acceleration = GetGeometry()[0].FastGetSolutionStepValue(VOLUME_ACCELERATION);
 
     // Compute and add external forces
-    const double Nodal_Mass = Element::GetValue(NODAL_MASS);
-    for ( unsigned int j = 0; j < dimension; j++ )
-    {
-        rRightHandSideVector[j]  += Volume_Acceleration[j] * Nodal_Mass;
+    if (this->Has(NODAL_MASS)){
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: you can do it nicer by greating a const reference to the element;
the important point is that if you call the "const" version of the method and the variable is not there a zero is given back without allocating something new in the database

in short what i propose doing is:
     
      const auto& rconst_this = *this;
      rconst_this.GetValue(NODAL_MASS);

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp [hunk 3]
```diff
@@ -300,36 +279,40 @@ void NodalConcentratedElement::CalculateRightHandSide(VectorType& rRightHandSide
...
+        std::cout << "WARNING :: YOUR NODAL ELEMENT ID: << this->Id() << LACKS OF NODAL MASS" << std::endl;
+#endif
 
     // Compute and add internal forces
-    const array_1d<double, 3 > Nodal_Stiffness = Element::GetValue(NODAL_STIFFNESS);
-    for ( unsigned int j = 0; j < dimension; j++ )
-    {
-        rRightHandSideVector[j]  -= Nodal_Stiffness[j] * Current_Displacement[j];
+    if (this->Has(NODAL_STIFFNESS)){
+        const array_1d<double, 3 >& nodal_stiffness = this->GetValue(NODAL_STIFFNESS);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: const array_1d<double,3>& nodal_stiffness = rconst_this.GetValue(NODAL_STIFFNESS)

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: I didn't know this kind of magic, but it has been probably the revelation of the year :heart_eyes:

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp (hunk 3 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: Marvelous

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp (hunk 1 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: Yep

**Comment by loumalouomega** [comment]
Ping

**Review by RiccardoRossi — COMMENTED** [review]
sE
======================================================================
ERROR: test_undamped_mdof_harmonic (__main__.HarmonicAnalysisTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_harmonic_analysis.py", line 129, in test_undamped_mdof_harmonic
    self._solve_eigen(mp)
  File "test_harmonic_analysis.py", line 69, in _solve_eigen
    eig_strategy.Solve()
RuntimeError: Error: Element 3 has non-positive size 0

Problem in the Check in the NodalConcentratedElement





in kratos/includes/element.h:883:virtual int Element::Check(const ProcessInfo&)
   kratos/includes/element.h:887:virtual int Element::Check(const ProcessInfo&)
   /home/rrossi/kratos_git/applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp:479:virtual int NodalConcentratedElement::Check(const ProcessInfo&)
   kratos/solving_strategies/strategies/solving_strategy.h:394:int SolvingStrategy<TSparseSpace, TDenseSpace, TLinearSolver>::Check() [with TSparseSpace = UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >; TDenseSpace = UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >; TLinearSolver = LinearSolver<UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >, UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > >]
   /home/rrossi/kratos_git/applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:519:int EigensolverStrategy<TSparseSpace, TDenseSpace, TLinearSolver>::Check() [with TSparseSpace = UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >; TDenseSpace = UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >; TLinearSolver = LinearSolver<UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >, UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > >]
   /home/rrossi/kratos_git/applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:280:void EigensolverStrategy<TSparseSpace, TDenseSpace, TLinearSolver>::Initialize() [with TSparseSpace = UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >; TDenseSpace = UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >; TLinearSolver = LinearSolver<UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >, UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > >]
   /home/rrossi/kratos_git/applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:348:double EigensolverStrategy<TSparseSpace, TDenseSpace, TLinearSolver>::Solve() [with TSparseSpace = UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >; TDenseSpace = UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >; TLinearSolver = LinearSolver<UblasSpace<double, boost::numeric::ublas::compressed_matrix<...>, boost::numeric::ublas::vector<double> >, UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > >]

**Comment by loumalouomega** [comment]
Are your running in full debug? I am running in full debug and I get an OK

**Comment by loumalouomega** [comment]
Ok, I discovered the problem, I need to compile ComplexPastix or it skips the test, sorry

**Comment by loumalouomega** [comment]
The check from the base element crashes the check (is a node, 0 size then). Sorry I don't use to compile Pastix in debug, so it skipped the tests

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp [hunk 4]
```diff
@@ -393,21 +364,17 @@ void NodalConcentratedElement::CalculateMassMatrix( MatrixType& rMassMatrix, Pro
...
-        rMassMatrix.resize( SystemSize, SystemSize, false );
-    }
+    if ( rMassMatrix.size1() != system_size )
+        rMassMatrix.resize( system_size, system_size, false );
 
-    rMassMatrix = ZeroMatrix( SystemSize, SystemSize );
+    rMassMatrix = ZeroMatrix( system_size, system_size );
 
-    double &Nodal_Mass = Element::GetValue(NODAL_MASS);
+    const double nodal_mass = this->GetValue(NODAL_MASS);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: note that putting "const double"  is not enough. you need to make const the reference you call it onto.

that is you need to do. Doing this the NODAL_MASS would still be allocated
    
      const auto& rmyself = *this;
      rmyself.GetValue(...)

**Review by RiccardoRossi — DISMISSED** [review]

**Comment by loumalouomega** [comment]
Ping

**Review by philbucher — APPROVED** [review]

**Pull request merged**